### PR TITLE
Expose delimiter character in JSON reader options to JSON reader APIs

### LINKED ISSUE #17261: [BUG] `cudf::io::json::detail::normalize_single_quotes` outputs incorrect result when the input has `\n` character
Reproducible with this input:
```
{\"a\": \"1\n2\"}
{\'a\': 12}
```

The output tokens, generated by `cudf::io::json::detail::get_token_stream` after preprocessing with `cudf::io::json::detail::normalize_single_quotes` are:
```
Input:
{"a": "1
2"}{'a': 12}
Tokens:
0, 4, 6, 7, 8, 9, 5, 1, 0, 1
Token indices:
0, 1, 1, 3, 6, 10, 11, 11, 0, 0
```

If remove the `\n` character then the output is correct:
```
Input:
{"a": "12"}{"a": 12}
Tokens:
0, 4, 6, 7, 8, 9, 5, 1, 0, 4, 6, 7, 10, 11, 5, 1
Token indices:
0, 1, 1, 3, 6, 9, 10, 10, 12, 13, 13, 15, 18, 20, 20, 20
```
Note: 
 * Line delimiter between JSON objects is `\0`, not `\n`.
 * `allow_unquoted_control` is set to `true`.
 * Token indices are the positions of the tokens in the input string.
 * Token numbers are `static_cast` from `enum token_t` at https://github.com/rapidsai/cudf/blob/06b3f83b3e7f1b1364973be34f58fac4caf773f3/cpp/include/cudf/io/detail/tokenize_json.hpp#L88

I suspect that it is due to the leftover character `\n` in https://github.com/rapidsai/cudf/blob/06b3f83b3e7f1b1364973be34f58fac4caf773f3/cpp/src/io/json/json_normalization.cu#L76-L77, but I'm not 100% sure.

### PR DESCRIPTION
**Pull request by shrshi**
## Description
Fixes #17261 
Removes delimiter symbol group from whitespace normalization FST since it is run post-tokenization.

## Checklist
- [X] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [X] New or existing tests cover these changes.
- [X] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/tests/io/json/json_quote_normalization_test.cpp [hunk 1]
```diff
@@ -179,22 +179,24 @@ TEST_F(JsonNormalizationTest, ReadJsonOption)
...
+  std::string const expected_input = R"({"a": "1\n2"}h{"a": 12})";
   cudf::io::json_reader_options expected_input_options =
     cudf::io::json_reader_options::builder(
       cudf::io::source_info{expected_input.data(), expected_input.size()})
-      .lines(true);
+      .lines(true)
+      .delimiter('h');
 
   cudf::io::table_with_metadata expected_table =
     cudf::io::read_json(expected_input_options, cudf::test::get_default_stream(), rsc.get());
```
Comment (karthikeyann) [inline cpp/tests/io/json/json_quote_normalization_test.cpp:202]: Please add one more test with a golden output. Right how it compares read_json against read_json result.
The existing output shows impact of `normalize_single_quotes`. Adding a test on the `normalize_single_quotes` API would be nice.

File: cpp/include/cudf/io/detail/json.hpp [hunk 2]
```diff
@@ -62,6 +62,7 @@ void write_json(data_sink* sink,
  * @param mr Device memory resource to use for device memory allocation
  */
 void normalize_single_quotes(datasource::owning_buffer<rmm::device_buffer>& indata,
+                             char delimiter,
```
Comment (ttnghia) [inline cpp/include/cudf/io/detail/json.hpp:65]: Need to update `@param` for this.

File: cpp/tests/io/json/json_quote_normalization_test.cpp (hunk 1 above)
Comment (shrshi) [reply cpp/tests/io/json/json_quote_normalization_test.cpp:211]: Added another test

**Review by ttnghia — APPROVED** [review]

File: cpp/tests/io/json/json_quote_normalization_test.cpp [hunk 3]
```diff
@@ -172,29 +174,38 @@ TEST_F(JsonNormalizationTest, GroundTruth_QuoteNormalization_Invalid_WrongBraces
   run_test(input, output);
 }
 
+TEST_F(JsonNormalizationTest, GroundTruth_QuoteNormalization_NonNewlineDelimiter)
+{
+  std::string input  = R"({'a': 'b\n'}z{'c\n"': "d\"")";
+  std::string output = R"({"a": "b\n"}z{"c\n\"": "d\"")";
+  run_test(input, output);
+}
+
```
Comment (karthikeyann) [inline cpp/tests/io/json/json_quote_normalization_test.cpp:183]: delimiter is not passed here.

Comment (shrshi) [reply cpp/tests/io/json/json_quote_normalization_test.cpp]: Oops, fixed now!

File: cpp/src/io/json/read_json.cu [hunk 4]
```diff
@@ -248,7 +248,8 @@ table_with_metadata read_batch(host_span<std::unique_ptr<datasource>> sources,
   // If input JSON buffer has single quotes and option to normalize single quotes is enabled,
   // invoke pre-processing FST
   if (reader_opts.is_enabled_normalize_single_quotes()) {
-    normalize_single_quotes(bufview, stream, cudf::get_current_device_resource_ref());
+    normalize_single_quotes(
+      bufview, reader_opts.get_delimiter(), stream, cudf::get_current_device_resource_ref());
```
Comment (davidwendt) [inline cpp/src/io/json/read_json.cu:252]: Just for verification. The `get_delimiter()` will return `\n` if the caller did not set a delimiter in the `json_reader_options`?

Comment (shrshi) [reply cpp/src/io/json/read_json.cu:252]: Yes, the default value set in `json_reader_options` for the delimiter is `\n` (https://github.com/rapidsai/cudf/blob/150d8d8c3a3a4aec87004fd0d130b56e388fa43d/cpp/include/cudf/io/json.hpp#L116)

**Review by davidwendt — APPROVED** [review]

**Review by karthikeyann — APPROVED** [review]

File: cpp/src/io/json/json_normalization.cu [hunk 5]
```diff
@@ -72,13 +72,18 @@ constexpr auto TT_SEC            = dfa_states::TT_SEC;
 constexpr auto TT_NUM_STATES     = static_cast<StateT>(dfa_states::TT_NUM_STATES);
 constexpr auto NUM_SYMBOL_GROUPS = static_cast<uint32_t>(dfa_symbol_group_id::NUM_SYMBOL_GROUPS);
 
-// The i-th string representing all the characters of a symbol group
-std::array<std::vector<SymbolT>, NUM_SYMBOL_GROUPS - 1> const qna_sgs{
-  {{'\"'}, {'\''}, {'\\'}, {'\n'}}};
+template <typename SymbolT>
+auto get_sgid_lut(SymbolT delim)
```
[the remark this answers is not in the document]
Comment (karthikeyann) [reply cpp/src/io/json/json_normalization.cu:76]:
```suggestion
auto get_sgid_lut(SymbolT delim)
```

**Pull request merged**